third_party/amd/lib/TritonAMDGPUTransforms/Utility.cpp

```diff
@@ -0,0 +1,96 @@
+#include "Utility.h"
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+namespace findMin {
+int findMinInBlock(Block &block,
+                   const std::function<int(Operation *)> &countFunc);
+
+// Returns the minimum found when accumulating countFunc(op) for all paths
+// between op1 and op2
+int findMinBetweenOps(Operation *op1, Operation *op2,
+                      const std::function<int(Operation *)> &countFunc) {
+  assert(op1 && op2);
+  int count = 0;
+  for (auto op = op1; op != op2; op = op->getNextNode()) {
+    if (auto ifOp = llvm::dyn_cast<scf::IfOp>(op)) {
+      assert(!ifOp.getThenRegion().empty() && !ifOp.getElseRegion().empty());
+      auto minThen = findMinInBlock(ifOp.getThenRegion().front(), countFunc);
+      auto minElse = findMinInBlock(ifOp.getElseRegion().front(), countFunc);
+      count += std::min(minThen, minElse);
+    } else if (auto forOp = llvm::dyn_cast<scf::ForOp>(op)) {
+      count += findMinInBlock(*forOp.getBody(), countFunc);
+    } else {
```

Comment:
Do we need to handle while op too?

Reply:
That part was actually incorrect. We can only count the instructions if we know the iteration count of the loop. Note that we only encounter a loop in this part of the code if the `async_wait` is not part of this loop so we have to execute the complete loop until we arrive at the producer/CommitGroup of the token.

I changed the code to multiply the count of the body with the iteration count if it's constant and ignore the loop otherwise. We also ignore whileOp because we cannot know its iteration count.